### PR TITLE
cvxcore optimization

### PR DESCRIPTION
**Pull request by akshayka**
Adds several optimizations to cvxcore, and adds parallelism.

This change makes the processing of `LinOp`s in cvxcore's `build_matrix` function happen in parallel, provided that cvxcore was compiled with OpenMP. The user can choose the number of threads to be used via `cvxpy.set_num_threads()`.

By default, cvxcore is _not_ built with OpenMP, since some systems might not have it. To build with OpenMP, specify the compiler/linker flags at the command line. eg,

```
CFLAGS="-fopenmp" LDFLAGS="-lgomp" python setup.py install
```

That said, we should make sure that our wheels are built with OpenMP by default. I'd appreciate it if someone could help with that, in a future PR.

Additionally, this change gets rid of some unnecessary copies in cvxcore.

### CODE REVIEW

File: cvxpy/cvxcore/src/LinOpOperations.cpp [hunk 1]
```diff
@@ -161,15 +161,19 @@ Tensor lin_to_tensor(const LinOp &lin) {
...
-Tensor build_tensor(const Matrix &mat) {
+Tensor build_tensor(Matrix &mat) {
   Tensor ten;
-  DictMat dm;
-  std::vector<Matrix> mat_vec;
-  mat_vec.push_back(mat);
-  dm[CONSTANT_ID] = mat_vec;
-  ten[CONSTANT_ID] = dm;
+  ten[CONSTANT_ID] = DictMat();
+  DictMat* dm = &(ten[CONSTANT_ID]);
```
Comment (SteveDiamond) [inline cvxpy/cvxcore/src/LinOpOperations.cpp:171]: Can you add a comment explaining what the difference is here?

**Comment by SteveDiamond** [comment]
Looks reasonable to me! I'd appreciate some comments explaining how you're doing things in a more copy-free/efficient way. Also looks like there are compile issues on travis.

**Comment by akshayka** [comment]
> Looks reasonable to me! I'd appreciate some comments explaining how you're doing things in a more copy-free/efficient way. Also looks like there are compile issues on travis.

Thanks for the review! Previously, the copy constructor of `SparseMatrix` was invoked implicitly in many places --- in the creation of temporaries, when pushing back an existing `DictMat` or `Matrix` into an STL container, when passing arguments by value, iterating over arguments by value, and so on. The change fixes many instances of things like that.

I'll fix the Travis CI.

File: cvxpy/cvxcore/src/LinOpOperations.cpp (hunk 1 above)
Comment (akshayka) [reply cvxpy/cvxcore/src/LinOpOperations.cpp:171]: Done.

File: cvxpy/cvxcore/src/cvxcore.cpp [hunk 2]
```diff
@@ -27,9 +33,14 @@
  * This function takes horizontal and vertical offset, which indicate
  * the offset of this block within our larger matrix.
  */
-void add_matrix_to_vectors(Matrix &block, std::vector<double> &V,
+void add_matrix_to_vectors(const Matrix &block, std::vector<double> &V,
                            std::vector<int> &I, std::vector<int> &J,
-                           int &vert_offset, int &horiz_offset) {
+                           int vert_offset, int horiz_offset) {
+  auto number_of_nonzeros = block.nonZeros();
```
Comment (rileyjmurray) [inline cvxpy/cvxcore/src/cvxcore.cpp:39]: Would this be the place where people could get OOM errors? If so, is there a graceful way to tell OpenMP that this step failed (and that it should wait before trying this part again)?

**Review by rileyjmurray — COMMENTED** [review]
Looks mostly good to me! I did ask one question in ``cvxcore.cpp`` about gracefully handling potential OOM errors. Other than that I'd say we need to make sure to update the web documentation. 

Edit: @SteveDiamond and @akshayka what do you think about adding in some tests (maybe with a flag on a certain TravisCI or AppVeyor build) to ensure multi-threaded cvxcore behaves as expected?

File: cvxpy/cvxcore/src/cvxcore.cpp (hunk 2 above)
Comment (akshayka) [reply cvxpy/cvxcore/src/cvxcore.cpp:39]: There's no way to do that, unfortunately.

However, what I thought to be OOM errors was in fact an unrelated bug.

**Comment by akshayka** [comment]
To add tests, the build would need to be configured to compile cvxcore with openmp. Do you have thoughts on how to do that cleanly?

**Comment by rileyjmurray** [comment]
The Travis config file can define an environment variable tailored to each build configuration. We'd need to edit the Travis installation script to check the value of that environment variable, and then proceed with or without OpenMP (as appropriate). We can make sure OpenMP on these systems just by installing from conda-forge. The one weird thing we'd need to do is explicitly set the number of OpenMP threads to 2 or 4 (https://docs.travis-ci.com/user/languages/c/#openmp-projects).

**Comment by akshayka** [comment]
Some timings: master, this branch 1 thread, this branch 8 threads (test_benchmarks.py).

This branch, single-threaded, is somewhat faster than master.

Most of these problems don't have many expression trees that need to be canonicalized, so you don't see a big difference in timings. However `diffcp_sdp` is 3x faster when using multiple threads.

master
```
cone_matrix_stuffing_with_many_constraints: avg=1.117e+00 s , std=0.000e+00 s (1 iterations)
.diffcp_sdp: avg=7.330e+00 s , std=0.000e+00 s (1 iterations)
.least_squares: avg=4.506e-03 s , std=0.000e+00 s (1 iterations)
.sConic canonicalization
(ECOS) solver time:  0.038621366
cvxpy time:  0.5479763150455322
Quadratic canonicalization
(OSQP) solver time:  0.020118558
cvxpy time:  0.48471370976123046
.qp: avg=5.823e-03 s , std=0.000e+00 s (1 iterations)
.small_cone_matrix_stuffing: avg=3.141e-02 s , std=5.636e-03 s (10 iterations)
.small_lp: avg=1.476e-02 s , std=0.000e+00 s (1 iterations)
small_lp_second_time: avg=1.024e-03 s , std=0.000e+00 s (1 iterations)
.small_parameterized_cone_matrix_stuffing: avg=1.550e+00 s , std=0.000e+00 s (1 iterations)
.small_parameterized_lp: avg=1.503e+00 s , std=0.000e+00 s (1 iterations)
small_parameterized_lp_second_time: avg=2.394e-03 s , std=0.000e+00 s (1 iterations)
.tv_inpainting: avg=2.106e+00 s , std=0.000e+00 s (1 iterations)
```

1 thread (total time 14.2 seconds)
```
cone_matrix_stuffing_with_many_constraints: avg=1.063e+00 s , std=0.000e+00 s (1 iterations)
.diffcp_sdp: avg=6.490e+00 s , std=0.000e+00 s (1 iterations)
.least_squares: avg=4.448e-03 s , std=0.000e+00 s (1 iterations)
.sConic canonicalization
(ECOS) solver time:  0.038415661
cvxpy time:  0.4627268408310547
Quadratic canonicalization
(OSQP) solver time:  0.020128029
cvxpy time:  0.44349930940942384
.qp: avg=5.738e-03 s , std=0.000e+00 s (1 iterations)
.small_cone_matrix_stuffing: avg=2.831e-02 s , std=6.154e-03 s (10 iterations)
.small_lp: avg=1.523e-02 s , std=0.000e+00 s (1 iterations)
small_lp_second_time: avg=1.071e-03 s , std=0.000e+00 s (1 iterations)
.small_parameterized_cone_matrix_stuffing: avg=1.313e+00 s , std=0.000e+00 s (1 iterations)
.small_parameterized_lp: avg=1.285e+00 s , std=0.000e+00 s (1 iterations)
small_parameterized_lp_second_time: avg=2.445e-03 s , std=0.000e+00 s (1 iterations)
.tv_inpainting: avg=1.919e+00 s , std=0.000e+00 s (1 iterations)
```

8 threads (total time 9.6 seconds)
```
cone_matrix_stuffing_with_many_constraints: avg=9.566e-01 s , std=0.000e+00 s (1 iterations)
.diffcp_sdp: avg=2.033e+00 s , std=0.000e+00 s (1 iterations)
.least_squares: avg=5.242e-03 s , std=0.000e+00 s (1 iterations)
.sConic canonicalization
(ECOS) solver time:  0.03815672
cvxpy time:  0.49952104489257815
Quadratic canonicalization
(OSQP) solver time:  0.020107715999999998
cvxpy time:  0.4717443290592041
.qp: avg=6.578e-03 s , std=0.000e+00 s (1 iterations)
.small_cone_matrix_stuffing: avg=3.314e-02 s , std=8.332e-03 s (10 iterations)
.small_lp: avg=2.008e-02 s , std=0.000e+00 s (1 iterations)
small_lp_second_time: avg=1.036e-03 s , std=0.000e+00 s (1 iterations)
.small_parameterized_cone_matrix_stuffing: avg=1.214e+00 s , std=0.000e+00 s (1 iterations)
.small_parameterized_lp: avg=1.367e+00 s , std=0.000e+00 s (1 iterations)
small_parameterized_lp_second_time: avg=2.414e-03 s , std=0.000e+00 s (1 iterations)
.tv_inpainting: avg=1.866e+00 s , std=0.000e+00 s (1 iterations)
```

**Comment by akshayka** [comment]
I made a few more small optimizations (eliminating copies), upgraded Eigen to 3.3.9, and switched to pruned sparse-sparse multiplication (which works in 3.3.9, but was broken in the version we were using).

There are still some copies that I wasn't able to get rid of. Most notably `get_constant_data` copies out numerical data instead of mapping it. But there are other copies as well. We can address those in a later change.

The improvements are fairly significant. See below.

master (total time 15.6s)
```
cone_matrix_stuffing_with_many_constraints: avg=1.122e+00 s , std=0.000e+00 s (1 iterations)
.diffcp_sdp: avg=7.410e+00 s , std=0.000e+00 s (1 iterations)
.least_squares: avg=4.661e-03 s , std=0.000e+00 s (1 iterations)
.sConic canonicalization
(ECOS) solver time:  0.039003118
cvxpy time:  0.5048685231590576
Quadratic canonicalization
(OSQP) solver time:  0.020439915
cvxpy time:  0.4658264595880127
.qp: avg=5.785e-03 s , std=0.000e+00 s (1 iterations)
.small_cone_matrix_stuffing: avg=2.912e-02 s , std=5.820e-03 s (10 iterations)
.small_lp: avg=1.465e-02 s , std=0.000e+00 s (1 iterations)
small_lp_second_time: avg=1.029e-03 s , std=0.000e+00 s (1 iterations)
.small_parameterized_cone_matrix_stuffing: avg=1.470e+00 s , std=0.000e+00 s (1 iterations)
.small_parameterized_lp: avg=1.450e+00 s , std=0.000e+00 s (1 iterations)
small_parameterized_lp_second_time: avg=2.410e-03 s , std=0.000e+00 s (1 iterations)
.tv_inpainting: avg=2.087e+00 s , std=0.000e+00 s (1 iterations)
```


this PR, 1 thread (total time 10.5s)
```
cone_matrix_stuffing_with_many_constraints: avg=8.530e-01 s , std=0.000e+00 s (1 iterations)
.diffcp_sdp: avg=3.764e+00 s , std=0.000e+00 s (1 iterations)
.least_squares: avg=4.784e-03 s , std=0.000e+00 s (1 iterations)
.sConic canonicalization
(ECOS) solver time:  0.039389905
cvxpy time:  0.392903271651001
Quadratic canonicalization
(OSQP) solver time:  0.020359027999999998
cvxpy time:  0.42363215623461914
.qp: avg=5.808e-03 s , std=0.000e+00 s (1 iterations)
.small_cone_matrix_stuffing: avg=2.845e-02 s , std=6.201e-03 s (10 iterations)
.small_lp: avg=1.412e-02 s , std=0.000e+00 s (1 iterations)
small_lp_second_time: avg=1.026e-03 s , std=0.000e+00 s (1 iterations)
.small_parameterized_cone_matrix_stuffing: avg=1.211e+00 s , std=0.000e+00 s (1 iterations)
.small_parameterized_lp: avg=1.122e+00 s , std=0.000e+00 s (1 iterations)
small_parameterized_lp_second_time: avg=2.473e-03 s , std=0.000e+00 s (1 iterations)
.tv_inpainting: avg=1.724e+00 s , std=0.000e+00 s (1 iterations)
```

this PR, 8 threads (total time 8.1s)
```
cone_matrix_stuffing_with_many_constraints: avg=7.508e-01 s , std=0.000e+00 s (1 iterations)
.diffcp_sdp: avg=1.510e+00 s , std=0.000e+00 s (1 iterations)
.least_squares: avg=5.197e-03 s , std=0.000e+00 s (1 iterations)
.sConic canonicalization
(ECOS) solver time:  0.038531578
cvxpy time:  0.4179091091795654
Quadratic canonicalization
(OSQP) solver time:  0.02024423
cvxpy time:  0.4240571368060303
.qp: avg=6.654e-03 s , std=0.000e+00 s (1 iterations)
.small_cone_matrix_stuffing: avg=3.045e-02 s , std=6.115e-03 s (10 iterations)
.small_lp: avg=1.897e-02 s , std=0.000e+00 s (1 iterations)
small_lp_second_time: avg=1.156e-03 s , std=0.000e+00 s (1 iterations)
.small_parameterized_cone_matrix_stuffing: avg=1.092e+00 s , std=0.000e+00 s (1 iterations)
.small_parameterized_lp: avg=1.097e+00 s , std=0.000e+00 s (1 iterations)
small_parameterized_lp_second_time: avg=2.428e-03 s , std=0.000e+00 s (1 iterations)
.tv_inpainting: avg=1.741e+00 s , std=0.000e+00 s (1 iterations)
```

**Comment by akshayka** [comment]
I believe this PR is almost done. There are two outstanding things.

1) Testing multithreaded compilation on Travis.

I've added an entry to the build matrix that builds and tests with OpenMP enabled. This works on linux, but is currently failing on macOS. @rileyjmurray or @SteveDiamond , can you help me get the macOS build passing? I'm not exactly sure what's wrong, because Travis is truncating the log. My guess is that I'm passing the wrong compiler/linker flags to gcc. I don't own a mac, so I can't test this locally.


2) Web documentation.

My vote is to add the web docs in a future PR. Right now we don't really say anything about performance on cvxpy.org. I am of the opinion that we should add a new section on performance. This section will emphasize the fact that CVXPY is a compiler, and will have tips on how to get better compilation performance, including compiling with OpenMP.

**Review by rileyjmurray — APPROVED** [review]
The number of lines changed jumped massively, but I trust this is just because of updating Eigen. Approved!

**Pull request merged**